test/stateChangeCapture.test.ts

```diff
@@ -108,21 +108,92 @@ describe('stateChangeCapture', () => {
         ['foo_my-pipeline1234345jkdk', 'foo'],
         ['foo_my_pipeline1234345jkdk', 'foo'],
         ['foo-this-doesnt-match-anything', 'foo-this-doesnt-match-anything'],
+        ['foo-pipeline', 'foo'],
+        ['foo-pipeline-123', 'foo'],
+        ['foo_pipeline', 'foo'],
+        ['foo_pipeline397654', 'foo'],
+      ];
+
+
+      it.each(examples)('should sanitise matching naming patterns in order of preference', (actual, expected) => {
+        expect(sanitizePipelineName(actual)).toEqual(expected);
+      })
+
+      it.each(examples)('should strip whitespace from SANITISE_PATTERNS variable', (actual, expected) => {
+        process.env.SANITISE_PATTERNS = "-codePipeline,_codePipeline,-my-pipeline,_my-pipeline,-my_pipeline,_my_pipeline,-pipeline,_pipeline"
+        expect(sanitizePipelineName(actual)).toEqual(expected);
+      })
+    })
+
+    describe('single regex based naming convention defined', () => {
+      beforeEach(() => {
+        process.env.IS_REGEX = "true"
+        process.env.SANITISE_PATTERNS = "(.*)-my-pipeline"
+      })
+
+      const examples = [
+        ['foo-pipeline', 'foo-pipeline'],
+        ['foo-pipeline-123', 'foo-pipeline-123'],
+        ['foo_pipeline', 'foo_pipeline'],
+        ['foo_pipeline397654', 'foo_pipeline397654'],
+        ['foo-codePipeline', 'foo-codePipeline'],
+        ['foo-codePipeline0987655', 'foo-codePipeline0987655'],
+        ['foo_codePipeline', 'foo_codePipeline'],
+        ['foo_codePipeline876tghujn', 'foo_codePipeline876tghujn'],
+        ['foo-my-pipeline', 'foo'],
+        ['foo-my-pipeline1234345jkdk', 'foo'],
+        ['foo_my-pipeline1234345jkdk', 'foo_my-pipeline1234345jkdk'],
+        ['foo_my_pipeline1234345jkdk', 'foo_my_pipeline1234345jkdk'],
+      ];
+
+      it.each(examples)('should sanitise matching naming patterns', (actual, expected) => {
+        expect(sanitizePipelineName(actual)).toEqual(expected);
+      })
+
+    })
+    describe('multiple regex pattern based naming conventions defined', () => {
+      beforeEach(() => {
+        process.env.IS_REGEX = "true"
+        process.env.SANITISE_PATTERNS = "[a-zA-z0-9]+(?=-codePipeline),(.*)(?=-codePipeline),(.*)(?=_codePipeline),(.*)(?=-my-pipeline),(.*)(?=_my-pipeline),(.*)(?=-my_pipeline),(.*)(?=_my_pipeline),(.*)(?=-pipeline),(.*)(?=_pipeline),(.*)(?=-codePipeline)  ,   (.*)(?=_codePipeline)   ,(.*)(?=-my-pipeline),(.*)(?=_my-pipeline), (.*)(?=-my_pipeline),(.*)(?=_my_pipeline),    (.*)(?=-pipeline)  ,  (.*)(?=_pipeline),(?<=foo-)(.*)+(?=@master)"
```

Comment:
Hmmm, sorry about this.

Reviewing again, I'm concerned that this would be overly complex to be really usable now that I am seeing the implications.

There are also a number of scenarios that aren't currently covered here.

Options:
* Full regular expression support
* Simple wildcard support (similar to your original version)
* simplified regex/placeholder syntax (similar to Cucumber expressions)


I'd like to investigate some examples:
In the examples below, by `complex` I am referring to generated/random content.

| Use Case                    | Example Pipeline Name | Expected sanitised name | Example Pattern | Match | 
|------------------------------|----------------------------------|-----------| ------------|----------|
| ignore common suffix    | foo-pipeline                     | foo       | -pipeline             | YES   | 
| ignore complex suffix    | foo-pipeline123abc          | foo       |-pipeline             | YES   | 
| ignore simple prefix      | pipeline-foo                      | foo       |                           | YES  | 
| ignore complex prefix   | pipeline123abc-foo          | foo       |                           | YES  | 
| Extract an infix             | pipeline123-foo-someSuffix | foo   |                           | YES  | 
| Crazy multi infix           | foo-abc123-bar-abc123 | something-important |  | NO |

Reply:
Considering the above, I can think of the following possible patterns:
```
*-pipeline
{capture}-pipeline
/pipeline.*-(.*)/
/([^-]*)-.*(-.*)-.*/
{word}-{anything}-{word}
```

Hmm not sure about any of those :thinking: 

Suggestions welcome

Reply:
Other thoughts on alternatives:

```
startsWith(pipeline)
```

Reply:
I also did think it would become overly complicated to be usable. It would rely on the end-user's  ability to provide Regex patterns they are searching for which can really range.
After discussing with Z, we thought that the actual end-user is the CNF team who will provide these patterns to match the end-users pipeline names.
Do we still want to do this? Shall i come up with the examples you've mentioned above as test cases and the examples would look for all regex matches and concatenate them / or output the first match as the sanitzed pattern?

Do you think it would be usable for other people using the SAR who might not be the CNF team? Do you see other people using this?